fixedpoint/fixedpoint_sse.h

```diff
@@ -0,0 +1,263 @@
+inline bool Any(__m128i a) {
+  return BitNot(_mm_testz_si128(a, a));
+}
+
+template <>
+inline __m128i RoundingHalfSum(__m128i a, __m128i b) {
+  /* __m128i round_bit_mask, a_over_2, b_over_2, round_bit, sum; */
+  /* We divide the inputs before the add to avoid the overflow and costly test */
+  /* of checking if an overflow occured on signed add */
+  /* round_bit_mask = _mm_set1_epi32(1); */
+  /* a_over_2 = _mm_srai_epi32(a, 1); */
+  /* b_over_2 = _mm_srai_epi32(b, 1); */
+  /* sum = Add(a_over_2, b_over_2); */
+  /* round_bit = _mm_sign_epi32(BitAnd(BitOr(a,b), round_bit_mask), sum); */
+  /* return Add(sum, round_bit); */
+
+  /* Other possibility detecting overflow and xor the sign if an overflow happened*/
+  __m128i one, sign_bit_mask, sum, rounded_half_sum, overflow, result;
+  one = _mm_set1_epi32(1);
+  sign_bit_mask = _mm_set1_epi32(0x80000000);
+  sum = Add(a,b);
+  rounded_half_sum = _mm_srai_epi32(Add(sum, one), 1);
+  overflow = BitAnd(BitAnd(BitXor(a,rounded_half_sum), BitXor(b,rounded_half_sum)), sign_bit_mask);
+  result = BitXor(rounded_half_sum, overflow);
+  return result;
+}
+
+template <>
+inline __m128i SaturatingRoundingDoublingHighMul(__m128i a, __m128i b) {
+  __m128i min, saturation_mask, a0_a2, a1_a3, b0_b2, b1_b3;
+  __m128i a0b0_a2b2, a1b1_a3b3, a0b0_a2b2_rounded, a1b1_a3b3_rounded;
+  __m128i a0b0_a2b2_rounded_2x, a1b1_a3b3_rounded_2x, result;
+  __m128i nudge;
+  
+
+  // saturation only happen if a == b == INT_MIN
+  min =  _mm_set1_epi32(std::numeric_limits<int32_t>::min());
+  saturation_mask = BitAnd(MaskIfEqual(a,b), MaskIfEqual(a, min));
+
+  //a = a0 | a1 | a2 | a3
+  //b = b0 | b1 | b2 | b3
+  a0_a2 = a;
+  a1_a3 = _mm_srli_si128(a, 4);
+  b0_b2 = b;
+  b1_b3 = _mm_srli_si128(b, 4);
+  
+  a0b0_a2b2 =_mm_mul_epi32(a0_a2, b0_b2);
+  a1b1_a3b3 =_mm_mul_epi32(a1_a3, b1_b3);
+
+  //do the rounding and take into account that it will be doubled, so add 1<<30 or (1 - (1 << 30)) if negative
+  nudge = _mm_set1_epi64x(1 << 30);
+  /* positive_nudge = _mm_set1_epi64x(1 << 30); */
+  /* negative_nudge = _mm_set1_epi64x((1 << 30)); */
+  
+  /* nudge_mask1 = _mm_cmpgt_epi64(_mm_set1_epi64x(0), a0b0_a2b2); */
+  /* nudge1 = SelectUsingMask(nudge_mask1, */
+  /* 			   negative_nudge, */
+  /* 			   positive_nudge); */
+  a0b0_a2b2_rounded = _mm_add_epi64(a0b0_a2b2, nudge);
+  
+  /* nudge_mask2 = _mm_cmpgt_epi64(_mm_set1_epi64x(0), a1b1_a3b3); */
+  /* nudge2 = SelectUsingMask(nudge_mask2, */
+  /* 			   negative_nudge, */
+  /* 			   positive_nudge); */
+  a1b1_a3b3_rounded = _mm_add_epi64(a1b1_a3b3, nudge);
+  
+  //do the doubling
+  a0b0_a2b2_rounded_2x = _mm_slli_epi64(a0b0_a2b2_rounded, 1);
+  a1b1_a3b3_rounded_2x = _mm_slli_epi64(a1b1_a3b3_rounded, 1);
+  
+  //getting the high part of the products
+  result = _mm_blend_epi16(_mm_srli_si128(a0b0_a2b2_rounded_2x, 4),
+			   a1b1_a3b3_rounded_2x,
+			   0xcc);
+  
+  //saturate those which overflowed
+  return SelectUsingMask(saturation_mask, min, result);
```

Comment:
2. Thanks, Good to know that 16bits has more extensive support than 32bit. Might tip the scales toward using 16bit for this stuff in the future.
3. OK, thanks for the answer. Please don't bother about it.